test/support/utils.rb

```diff
@@ -32,26 +32,59 @@ def debuggee_backlog
       "\n[DEBUGGEE SESSION LOG]\n> #{backlog.join('> ')}"
     end
 
+    TestInfo = Struct.new(:queue, :remote_debuggee_info, :mode, :backlog, :last_backlog, :internal_info)
+
     # This method will execute both local and remote mode by default.
     def debug_code(program, boot_options: '-r debug/start', remote: true, &block)
       check_line_num!(program)
 
-      @scenario = block
       write_temp_file(strip_line_num(program))
+      @scenario = []
+      block.call
+      @scenario.freeze
       inject_lib_to_load_path
 
       ENV['RUBY_DEBUG_NO_COLOR'] = 'true'
       ENV['RUBY_DEBUG_TEST_MODE'] = 'true'
 
-      debug_on_local boot_options
-
       if remote && !NO_REMOTE
-        debug_on_unix_domain_socket
-        debug_on_tcpip
+        th = [new_thread { debug_on_local boot_options, TestInfo.new(new_queue) },
+              new_thread { debug_on_unix_domain_socket TestInfo.new(new_queue) },
+              new_thread { debug_on_tcpip TestInfo.new(new_queue) }]
+        th.each do |t|
+          if fail_msg = t.join.value
+            th.each(&:kill)
+            flunk fail_msg
+          end
+        end
+      else
+        debug_on_local boot_options, TestInfo.new(new_queue)
       end
+      assert true
     end
 
-    ASK_CMD = %w[quit q delete del kill undisplay]
+    def new_thread &block
+      Thread.new do
```

Comment:
If thread terminates with an exception, `t.join` on line 55 will raise an exception.
It should rescue all exceptions and return it as a failure reason.

Reply:
https://github.com/ruby/debug/pull/180/files#diff-ee14bf65f14927f7778a246f90c5cec0c581be3e2658c2a8a3726c59db5be3f9R61-R64